sky/backends/onprem_utils.py

```diff
@@ -174,29 +175,43 @@ def check_local_installation(ips: List[str], auth_config: Dict[str, str]):
     ssh_credentials = (ssh_user, ssh_key, 'sky-admin-deploy')
     runners = command_runner.SSHCommandRunner.make_runner_list(
         ips, *ssh_credentials)
+    sky_ray_version = constants.SKY_REMOTE_RAY_VERSION
 
-    def _check_dependencies(runner: command_runner.SSHCommandRunner) -> None:
-        # Checks for global python3 installation.
+    def _install_and_check_dependencies(
+            runner: command_runner.SSHCommandRunner) -> None:
+        # Checks for python3 installation.
         backend_utils.run_command_and_handle_ssh_failure(
-            runner,
-            'python3 --version',
+            runner, ('python3 --version'),
             failure_message=f'Python3 is not installed on {runner.ip}.')
 
-        # Checks for global Ray installation (accessible by all users).
+        # Checks for pip3 installation.
         backend_utils.run_command_and_handle_ssh_failure(
-            runner,
-            'ray --version',
+            runner, ('pip3 --version'),
+            failure_message=f'Pip3 is not installed on {runner.ip}.')
+
+        # If Ray does not exist, installs Ray.
+        backend_utils.run_command_and_handle_ssh_failure(
+            runner, (f'ray --version || '
+                     f'(pip3 install ray[default]=={sky_ray_version})'),
             failure_message=f'Ray is not installed on {runner.ip}.')
 
-        # Checks for global Sky installation (accessible by all users). When
-        # Sky's job submission code is ran on a user's account, Sky calls the
-        # Ray cluster to prepare the user's job. Due to Ray's limitations,
-        # this is ran under the admin's environment, which requires Sky to be
-        # installed globally.
-        # TODO(mluo): Make Sky admin only.
+        # If Ray exists, check Ray version. If the version does not match
+        # raise an error.
+        backend_utils.run_command_and_handle_ssh_failure(
+            runner,
+            f'ray --version | grep {sky_ray_version}',
+            failure_message=(
+                f'Ray (on {runner.ip}) does not match skypilot\'s'
+                f' requirement for ray=={sky_ray_version}.'
+                f' Make sure that the correct version of ray is installed.'))
+
+        # Checks for Sky installation. When Sky's job submission code is ran
+        # on a user's account, Sky calls the Ray cluster to prepare the user's
+        # job. Due to Ray's limitations, this is ran under the admin's
+        # environment, which requires Sky to be installed globally.
         backend_utils.run_command_and_handle_ssh_failure(
             runner,
-            'sky --help',
+            'sky --help || (pip3 install skypilot)',
```

Comment:
Important note: skypilot 0.1.2 (our latest PyPI package as of 10/9/22) is broken - it does not contain #1209, and thus this pip install results in a `ModuleNotFoundError: No module named 'googleapiclient'`.

I was wondering if we should upload local SkyPilot source and install from that, but decided against it because releases are more likely to contain stable code. 

Action item: can we add a note here mentioning this detail? something like `NOTE: This package is installed from PyPI and may not contain any changes made since the last SkyPilot release. If required, please install skypilot from source on the onprem machine before running sky admin deploy`

Reply:
Seems to work with most updated `origin/master` ...lmk if you observe anything else otherwise!

I am thinking it is probably better to upload local Skypilot source. Sky Onprem will be a rapidly changing feature in the next couple weeks and the sparse version updates cannot keep up with it...

Regarding action item, done thanks!